core/src/main/java/com/datastax/oss/driver/internal/core/type/codec/registry/CachingCodecRegistry.java

```diff
@@ -68,19 +69,71 @@
 
   protected final String logPrefix;
   private final TypeCodec<?>[] primitiveCodecs;
-  private final TypeCodec<?>[] userCodecs;
+  private final CopyOnWriteArrayList<TypeCodec<?>> userCodecs = new CopyOnWriteArrayList<>();
   private final IntMap<TypeCodec> primitiveCodecsByCode;
 
   protected CachingCodecRegistry(
-      @NonNull String logPrefix,
-      @NonNull TypeCodec<?>[] primitiveCodecs,
-      @NonNull TypeCodec<?>[] userCodecs) {
+      @NonNull String logPrefix, @NonNull TypeCodec<?>[] primitiveCodecs) {
     this.logPrefix = logPrefix;
     this.primitiveCodecs = primitiveCodecs;
-    this.userCodecs = userCodecs;
     this.primitiveCodecsByCode = sortByProtocolCode(primitiveCodecs);
   }
 
+  /**
+   * @deprecated this constructor calls an overridable method ({@link #register(TypeCodec[])}),
+   *     which is a bad practice. The recommended alternative is to use {@link
+   *     #CachingCodecRegistry(String, TypeCodec[])}, then add the codecs with one of the {@link
+   *     #register} methods.
+   */
+  @Deprecated
+  protected CachingCodecRegistry(
+      @NonNull String logPrefix,
+      @NonNull TypeCodec<?>[] primitiveCodecs,
+      @NonNull TypeCodec<?>[] userCodecs) {
+    this(logPrefix, primitiveCodecs);
+    register(userCodecs);
+  }
+
+  @Override
+  public void register(TypeCodec<?> newCodec) {
+    for (TypeCodec<?> primitiveCodec : primitiveCodecs) {
+      if (primitiveCodec.accepts(newCodec.getCqlType())
+          && primitiveCodec.accepts(newCodec.getJavaType())) {
+        LOG.warn(
+            "[{}] Ignoring codec {} because it collides with built-in primitive codec {}",
+            logPrefix,
+            newCodec,
+            primitiveCodec);
+        return;
+      }
+    }
+    for (TypeCodec<?> userCodec : userCodecs) {
+      if (userCodec.accepts(newCodec.getCqlType()) && userCodec.accepts(newCodec.getJavaType())) {
+        LOG.warn(
+            "[{}] Ignoring codec {} because it collides with previously registered codec {}",
+            logPrefix,
+            newCodec,
+            userCodec);
+        return;
+      }
+    }
+    // Technically this would cover the two previous cases as well, but we want precise messages.
+    try {
+      TypeCodec<?> cachedCodec =
+          getCachedCodec(newCodec.getCqlType(), newCodec.getJavaType(), false);
+      LOG.warn(
+          "[{}] Ignoring codec {} because it collides with previously generated codec {}",
+          logPrefix,
+          newCodec,
+          cachedCodec);
+      return;
+    } catch (CodecNotFoundException ignored) {
+      // Catching the exception is ugly, but it avoids breaking the internal API (e.g. by adding a
```

Comment:
maybe we should log a warning?

Reply:
It would be overkill imo because we are only testing if such a codec has been generated; the nominal case would be that such a code should not exist, so the exception is actually the expected outcome.